### PR TITLE
"dune describe", for ROTOR and more

### PR DESCRIPTION
**Pull request by ghost**
This PR adds a `dune describe` command to extract various information about a workspace from dune for use by third party programs such as ROTOR. The query language and output format are versioned in the same way as dune configuration files.

The output can be in OCaml or canonical s-expression format.

This PR only describes library and not executables. I'm planning to add executables in a sub-sequent PR, but the current PR should be enough to get started on the ROTOR side.

@emillon, this is quite similar to your `dune describe-project` PR. I'd like to get this one merged soon as it is immediately useful for ROTOR, but after we should probably integrate your work into this command. In particular the nice terminal output.

## example

```
$ dune describe workspace --lang 2.3 --format ocaml
[ library
    { name = "foo"
    ; uid = "0c8684c86d6ffbcecee09205033aa0a5"
    ; requires = [ "da24c54ea66d10c3bf80362265e8accb" ]
    ; source_dir = "_build/default"
    ; modules =
        [ { name = "Foo"
          ; impl = Some "_build/default/foo.ml"
          ; intf = None
          ; cmt = Some "_build/default/.foo.objs/byte/foo.cmt"
          ; cmti = None
          }
        ]
    }
; library
    { name = "foo.x"
    ; uid = "da24c54ea66d10c3bf80362265e8accb"
    ; requires = []
    ; source_dir = "_build/default"
    ; modules =
        [ { name = "Foo_x"
          ; impl = Some "_build/default/foo_x.ml"
          ; intf = None
          ; cmt = Some "_build/default/.foo_x.objs/byte/foo_x.cmt"
          ; cmti = None
          }
        ]
    }
]
```

### CODE REVIEW

**Comment by diml** [comment]
Invoked via:

```
$ dune compute describe-everything default
```

**Comment by emillon** [comment]
Yes, I suppose we'll have to bite the bullet, it's been a while. I like the idea of passing an explicit `--lang`.

I'm a bit skeptical about the "ocaml" output. Is this required for ROTOR? What types does it correspond to?

**Comment by emillon** [comment]
(I'm of course happy to backport the terminal output to this API.)

**Comment by diml** [comment]
> I'm a bit skeptical about the "ocaml" output. Is this required for ROTOR? What types does it correspond to?

It's just because it was available (It's `Dyn.pp` basically). But yh, maybe it's a bit random, I changed it to be sexp-output by default.

I also changed things so that for now only `--lang 0.1` is available. Just so that we don't over-commit eagerly and we can tweak the output a bit. I expect that we will need to do that for ROTOR.

File: bin/describe.ml [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  ]
+
+let info = Term.info "describe" ~doc ~man
+
+(* Crawl the workspace to get all the data *)
+module Crawl = struct
+  open Dune
+
+  let uid_of_library lib =
+    Digest.generic (Lib.name lib, Lib_info.src_dir (Lib.info lib))
```
Comment (rgrinberg) [inline bin/describe.ml]: To be extra careful, maybe it's best to convert paths to strings here.

Comment (diml) [reply bin/describe.ml]: Sounds good, done

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**